components/prediction-page/prediction-page.tsx

```diff
@@ -1,91 +1,321 @@
+}
+
+function usePollingPrediction$(
+  { interval = 2000, pollId }: { interval?: number; pollId: string },
+) {
+  const prediction$ = useSignal<Poll>(undefined!);
+
+  const { signal$: predictionResponse$, reexecuteQuery: reexecutePredictionQuery } =
+    useUrqlQuerySignal(
+      POLL_QUERY,
+      {
+        id: pollId,
+      },
+    );
+
+  useEffect(() => {
+    const id = setInterval(() => {
+      reexecutePredictionQuery({ requestPolicy: "network-only" });
+    }, interval);
+
+    return () => clearInterval(id);
+  }, []);
+
+  // we use this to cut down rerenders in the prediction component by only updating the observable
+  // if the response has changed vs. on every interval
+  useObserve(() => {
+    const currentPrediction = predictionResponse$.data?.get()?.poll;
+    const pastPrediction = prediction$.get();
+
+    // only update if the prediction has changed
+    if (currentPrediction && !_.isEqual(currentPrediction, pastPrediction)) {
+      prediction$.set(currentPrediction);
     }
-  }
-`;
+  });
+
+  return { prediction$, predictionResponse$, reexecutePredictionQuery };
+}
+
+function usePollingActivePrediction$(
+  { interval = 2000 }: { interval?: number },
+) {
+  const prediction$ = useSignal<Poll>(undefined!);
+  const { signal$: predictionConnection$, reexecuteQuery: reexecutePredictionQuery } =
+    useUrqlQuerySignal(
+      ACTIVE_PREDICTION_QUERY,
+    );
+
+  useEffect(() => {
+    const id = setInterval(() => {
+      reexecutePredictionQuery({ requestPolicy: "network-only" });
+    }, interval);
+
+    return () => clearInterval(id);
+  }, []);
 
-const POLL_INTERVAL = 1000;
+  // we use this to cut down rerenders in the prediction component by only updating the observable
+  // if the response has changed vs. on every interval
+  useObserve(() => {
+    const activePrediction = predictionConnection$.data?.get()?.pollConnection.nodes[0];
+    const pastPrediction = prediction$.get();
 
-export default function PredictionPage() {
+    // only update if the prediction has changed
+    if (activePrediction && !_.isEqual(activePrediction, pastPrediction)) {
+      prediction$.set(activePrediction);
+    }
+  });
+
+  return { predictionConnection$, prediction$, reexecutePredictionQuery };
+}
+
+export default function PredictionPage({ pollId }: { pollId?: string }) {
   useStyleSheet(styleSheet);
-  const { data: channelPointsData } = usePollingQuery(POLL_INTERVAL, {
-    query: CHANNEL_POINTS_QUERY,
+
+  return pollId ? <PredictionByIdPage pollId={pollId} /> : <ActivePredictionPage />;
+}
+
+function ActivePredictionPage() {
+  const { prediction$, reexecutePredictionQuery, predictionConnection$ } =
+    usePollingActivePrediction$({ interval: 1000 });
+  const hasFetched$ = useSignal(false);
+  const isFetching$ = useSignal(false);
+
+  // we use this to ensure we only render the empty state if the prediction page has already fetched data
+  useObserve(() => {
+    if (predictionConnection$.fetching.get()) {
```

Comment:
don't feel super strongly, but this reads better to me imo

```
const isFetching = predictionConnection$.fetching.get()
isFetching$.set(isFetching)
if (isFetching && !hasFetched$.get()) {
  hasFetched.set(true)
}
```

really feel like legend should have some sort of distinctUntilChanged equivalent, which would avoid having to do the conditional too

Reply:
updated